consensus/safety-rules/src/persistent_safety_storage.rs

```diff
@@ -80,35 +85,56 @@ impl PersistentSafetyStorage {
         Ok(())
     }
 
+    pub fn into_cached(self) -> PersistentSafetyStorage {
+        match self.internal_store {
+            StorageVariant::Storage(storage) => PersistentSafetyStorage {
+                internal_store: StorageVariant::CachedStorage(storage.into_cached()),
+            },
+            StorageVariant::CachedStorage(storage) => PersistentSafetyStorage {
+                // idempotent
+                internal_store: StorageVariant::CachedStorage(storage),
+            },
+        }
+    }
+
     /// Use this to instantiate a PersistentStorage with an existing data store. This is intended
     /// for constructed environments.
     pub fn new(internal_store: Storage) -> Self {
-        Self { internal_store }
+        Self {
+            internal_store: StorageVariant::Storage(internal_store),
+        }
     }
 
-    pub fn author(&self) -> Result<Author> {
-        let res = self.internal_store.get(OPERATOR_ACCOUNT)?;
-        let res = res.value.string()?;
-        std::str::FromStr::from_str(&res)
+    pub fn author(&mut self) -> Result<Author> {
+        std::str::FromStr::from_str(&self.internal_store.get(OPERATOR_ACCOUNT)?.value.string()?)
     }
 
     pub fn consensus_key_for_version(
         &self,
         version: Ed25519PublicKey,
     ) -> Result<Ed25519PrivateKey> {
-        self.internal_store
-            .export_private_key_for_version(CONSENSUS_KEY, version)
-            .map_err(|e| e.into())
+        match &self.internal_store {
+            StorageVariant::Storage(storage) => storage
+                .export_private_key_for_version(CONSENSUS_KEY, version)
+                .map_err(|e| e.into()),
+            StorageVariant::CachedStorage(_) => {
+                Err(Error::InternalError("Cannot operate on keys with CachedStorage".into()).into())
+            }
+        }
```

Comment:
if we're going with this style, the cached layer would just pass any CryptoKV calls to the underlying storage